src/snowflake/snowpark/_internal/compiler/utils.py

```diff
@@ -103,49 +107,113 @@ def to_selectable(plan: LogicalPlan, analyzer: Analyzer) -> Selectable:
         for operand in parent.set_operands:
             if operand.selectable == old_child:
                 operand.selectable = new_child_as_selectable
-        return
 
-    if isinstance(parent, Selectable):
-        assert parent.snowflake_plan.source_plan is not None
-        replace_child(parent.snowflake_plan.source_plan, old_child, new_child, analyzer)
-        return
+    elif isinstance(parent, Selectable):
+        assert parent.snowflake_plan is not None
+        replace_child_and_reset_node(
+            parent.snowflake_plan, old_child, new_child, analyzer
+        )
 
-    if isinstance(parent, (UnaryNode, Limit, CopyIntoLocationNode)):
+    elif isinstance(parent, (UnaryNode, Limit, CopyIntoLocationNode)):
         parent.children = [new_child]
         parent.child = new_child
-        return
 
-    if isinstance(parent, BinaryNode):
+    elif isinstance(parent, BinaryNode):
         parent.children = [
             node if node != old_child else new_child for node in parent.children
         ]
         if parent.left == old_child:
             parent.left = new_child
         if parent.right == old_child:
             parent.right = new_child
-        return
 
-    if isinstance(parent, SnowflakeCreateTable):
+    elif isinstance(parent, SnowflakeCreateTable):
         parent.children = [new_child]
         parent.query = new_child
-        return
 
-    if isinstance(parent, (TableUpdate, TableDelete)):
+    elif isinstance(parent, (TableUpdate, TableDelete)):
         snowflake_plan = analyzer.resolve(new_child)
         parent.children = [snowflake_plan]
         parent.source_data = snowflake_plan
-        return
 
-    if isinstance(parent, TableMerge):
+    elif isinstance(parent, TableMerge):
         snowflake_plan = analyzer.resolve(new_child)
         parent.children = [snowflake_plan]
         parent.source = snowflake_plan
-        return
 
-    if isinstance(parent, LogicalPlan):
+    elif isinstance(parent, LogicalPlan):
         parent.children = [
             node if node != old_child else new_child for node in parent.children
         ]
-        return
 
-    raise ValueError(f"parent type {type(parent)} not supported")
+    else:
+        raise ValueError(f"parent type {type(parent)} not supported")
+
+    # reset the parent node to make sure it can be re-resolved in the later
+    # stage to get the correct query.
+    # For example with the following plan
+    #          SelectSnowflakePlan
+    #                  |
+    #             SnowflakePlan
+    #                  |
+    #                JOIN
+    # when reset the child of join, and parent is SelectSnowflakePlan, we make sure
+    # all nodes along the way up to SelectSnowflakePlan are reset correctly.
+    reset_node(parent)
+
+
+def reset_node(node: LogicalPlan) -> None:
+    """
+    Helper methods that is used to clean up cached query filed of the given node
```

Comment:
```suggestion
    Helper method that is used to clean up cached query filed of the given node
```
what's cached_query?